### PR TITLE
fix(object): #5736 — make own_key_present O(1) on wide objects so Object.values/entries aren't O(n²)

### LINKED ISSUE #5736: Per-module init cost at binary startup scales linearly with module count (~3.4 ms/module)
Building [`nub cook`](https://github.com/nubjs/nub/pull/162) (nubjs/nub#162) — a command that wraps `perry compile` to produce a native binary, with a Node fallback — I benchmarked the compiled binary's cold start against Node's startup options. `perry compile` wins decisively on small self-contained scripts, but the compiled binary's cold start grows linearly with the number of imported modules — about **3.4 ms per module** on this build. On a small self-contained script it starts in ~13 ms; on a 60-module graph it's ~207 ms. Reporting it here since it's a perry-side cost — I'm trying to understand whether this per-module init is expected and whether it can be reduced.

### Numbers

Module-count scaling sweep (K modules, 8 trivial exported functions each, a tiny startup-dominated driver). Each row is `hyperfine --warmup 5 --min-runs 40` cold start. Plain `node` on the same `.mts` source is the contrast.

| K (modules) | functions | perry compile | node |
| --- | --- | --- | --- |
| 1   | 8   | 13.1 ms ± 1.3  | 53.6 ms ± 28.1 |
| 5   | 40  | 18.5 ms ± 3.1  | 53.9 ms ± 2.3  |
| 15  | 120 | 58.8 ms ± 1.4  | 59.8 ms ± 2.5  |
| 30  | 240 | 115.4 ms ± 2.0 | 62.2 ms ± 1.9  |
| 45  | 360 | 167.5 ms ± 1.8 | 71.0 ms ± 1.8  |
| 60  | 480 | 207.3 ms ± 4.4 | 83.9 ms ± 3.0  |

Least-squares fit on the per-K means:

- **perry compile** = 7.6 ms floor + **3.43 ms/module** (R² = 0.996) — strongly linear in module count.
- **node** = 51 ms floor + **0.49 ms/module** (R² = 0.95).
- **Crossover** at **K ≈ 15 modules** (~120 functions): below it the compiled binary wins, above it the per-module term dominates and node pulls ahead.

The compiled floor (~8 ms) is the appealing part — a no-import script starts ~5× faster than node. The cost is entirely in the per-module slope: it's ~7× node's, so each added module is ~3 ms of binary-startup time rather than node's ~0.5 ms.

### Minimal repro

`gen.mjs` — emits K modules, a barrel, and a startup-dominated entry:

```js
import { mkdirSync, writeFileSync } from "node:fs";
import { join } from "node:path";
const K = Number(process.argv[2]), outdir = process.argv[3], F = 8, STEPS = 6;
mkdirSync(join(outdir, "mods"), { recursive: true });
const fn = (i, f) => {
  let s = `export function m${i}_f${f}(x: number): number {\n  let acc = x + ${i} * ${f};\n`;
  for (let k = 0; k < STEPS; k++) s += `  acc = (acc * ${k + 2} + ${i + f + k}) % 1000003;\n`;
  return s + `  return acc;\n}\n`;
};
for (let i = 0; i < K; i++) {
  let b = "";
  for (let f = 0; f < F; f++) b += fn(i, f);
  writeFileSync(join(outdir, "mods", `m${i}.mts`), b);
}
let barrel = "";
for (let i = 0; i < K; i++) barrel += `export * from "./m${i}.mts";\n`;
writeFileSync(join(outdir, "mods", "index.mts"), barrel);
writeFileSync(join(outdir, "entry.mts"), `import * as mods from "./mods/index.mts";
const fns = Object.values(mods).filter(v => typeof v === "function");
let acc = 0;
for (let i = 0; i < 100; i++) acc = (acc + fns[i % fns.length](i)) % 1000003;
console.log(acc);
`);
```

```sh
for K in 1 5 15 30 45 60; do
  d=$(mktemp -d); node gen.mjs "$K" "$d"
  perry compile "$d/entry.mts" -o "$d/cooked"
  echo "K=$K"; hyperfine --warmup 5 --min-runs 40 -N "$d/cooked"
done
```

### Environment

- perry 0.5.1189
- node v22.21.1 (`.mts` source compiled natively; native type-stripping for the node contrast)
- macOS, arm64

### Question

Is per-module initialization at binary startup expected to scale this way (every module's top-level initialized eagerly at boot), and is there a way to reduce it — e.g. lazy/on-first-use module init, or coalescing module init so the slope is closer to flat? The modules here are trivial (pure function exports, no top-level side effects), so I'd have expected init to be near-free; the ~3.4 ms/module suggests fixed per-module overhead rather than work proportional to module contents.

### PR DESCRIPTION
**Pull request by proggeramlug**
## Problem

Fixes #5736 (reported by @jdalton). Compiled-binary cold start appeared to scale ~linearly at **~3.4 ms/module**. Profiling the repro shows the cost is **not** module init — 240 trivial modules behind a barrel initialize in ~5 ms (~0.02 ms/module). The cost is the entry's `Object.values(mods)` over the `export *` **barrel namespace**, which was **O(modules²)**.

## Root cause

`Object.values` / `Object.entries` re-check every own key via `own_key_present` (an earlier value read can fire a getter that deletes a later key, so the key set is re-validated per element). `own_key_present` did an **O(n) linear scan** of the object's `keys_array`. For a star-export namespace with thousands of bindings, that's O(n) × n = **O(n²)**.

Isolation confirms it:
- `Object.keys(ns)` (names only) — **flat** (K=240: 10.4 ms)
- `Object.values(ns)` (resolves each value → per-key `own_key_present`) — **O(n²)** (K=240: 140 ms)
- barrel imported but *not* enumerated — 5.5 ms (modules init fine)

## Fix

The read-path getter already builds an O(1) key→slot index for wide objects (≥257 keys, #5054). `own_key_present` simply wasn't using it. Probe that index first; a hit is O(1), a **miss falls through to the existing linear scan**, so an absent key (or a present key whose index entry was dropped as stale) is still answered correctly — the index revalidates every hit against the live slot via `js_string_key_matches`. The change is a **no-op below the 257-key threshold**, so ordinary objects are unaffected.

## Validation

Repro (`gen.mjs` from the issue: K trivial modules, a barrel, entry does `Object.values(mods).filter(...)`):

| K | before | after |
| --- | --- | --- |
| 60 | 16.9 ms | 9.8 ms |
| 120 | 42.0 ms | 10.7 ms |
| 240 | **140.1 ms** | **12.3 ms** |

Per-module slope: **~3.4 ms/module → ~0** (flat). Output is identical before/after.

- Full `perry-runtime object::` test suite passes (the lone `closure_name_and_length_ignore_plain_assignment` blip is the repo's known parallel-execution side-table flake — passes in isolation, and is on a path my change doesn't touch).
- Adds `wide_object_own_key_present_uses_index_and_object_values_is_complete`: on a 600-key object, asserts present keys resolve, absent keys are rejected, and `Object.values` enumerates every value exactly once.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved performance of property existence checks for large objects by using a faster key lookup path first, while preserving exact behavior via a safe fallback.
  * Strengthened correctness of `Object.values` for large objects to ensure all expected values are returned with no duplicates or omissions, including boundary cases.
* **Tests**
  * Added a regression test for wide objects (600 keys) covering `ownKeyPresent` and `Object.values`, including present, absent, malformed, and edge-case keys.

### CODE REVIEW

File: crates/perry-runtime/src/object/tests.rs [hunk 1]
```diff
@@ -759,6 +759,55 @@ fn wide_object_index_reads_and_descriptor_writes() {
...
+        );
+        let mut sum = 0.0;
+        for i in 0..n {
+            let v = crate::array::js_array_get(values, i);
+            sum += f64::from_bits(v.bits());
+        }
+        // 0 + 1 + … + 599 = 599*600/2 = 179700
+        assert_eq!(
+            sum, 179_700.0,
+            "Object.values must yield exactly the set values"
```
[the remark this answers is not in the document]
Comment (proggeramlug) [reply crates/perry-runtime/src/object/tests.rs]: Good catch — applied in the latest commit. Replaced the length+sum check with a `seen` bitmap that asserts every value 0..n appears exactly once, so a balanced duplicate/omission can't slip through. Thanks!

**Comment by jdalton** [comment]
This seems to fix the module issue (https://github.com/PerryTS/perry/issues/5736 and https://github.com/nubjs/nub/pull/162):

>nub cook
